src/OrleansRuntime/Core/GrainMethodInvoker.cs

```diff
@@ -0,0 +1,182 @@
+        /// <param name="filters">The invocation interceptors.</param>
+        /// <param name="interfaceToImplementationMapping">The implementation map.</param>
+        /// <param name="invokeInterceptor">The deprecated silo-wide interceptor.</param>
+        public GrainMethodInvoker(
+            IAddressable grain,
+            InvokeMethodRequest request,
+            IGrainMethodInvoker rootInvoker,
+            List<IGrainCallFilter> filters,
+            InterfaceToImplementationMappingCache interfaceToImplementationMapping,
+            InvokeInterceptor invokeInterceptor)
+        {
+            this.request = request;
+            this.rootInvoker = rootInvoker;
+            this.Grain = grain;
+            this.filters = filters;
+            this.deprecatedInvokeInterceptor = invokeInterceptor;
+            this.interfaceToImplementationMapping = interfaceToImplementationMapping;
+        }
+
+        /// <inheritdoc />
+        public IAddressable Grain { get; }
+
+        /// <inheritdoc />
+        public MethodInfo Method
+        {
+            get
+            {
+                // Determine if the object being invoked is a grain or a grain extension.
+                Type implementationType;
+                var extensionMap = this.rootInvoker as IGrainExtensionMap;
+                IGrainExtension extension;
+                if (extensionMap != null && extensionMap.TryGetExtension(request.InterfaceId, out extension))
+                {
+                    implementationType = extension.GetType();
+                }
+                else
+                {
+                    implementationType = this.Grain.GetType();
+                }
+
+                // Get or create the implementation map for this object.
+                var implementationMap = interfaceToImplementationMapping.GetOrCreate(
+                    implementationType,
+                    request.InterfaceId);
+
+                // Get the method info for the method being invoked.
+                MethodInfo methodInfo;
+                implementationMap.TryGetValue(request.MethodId, out methodInfo);
+                return methodInfo;
+            }
+        }
+
+        /// <inheritdoc />
+        public object[] Arguments => request.Arguments;
+        
+        /// <inheritdoc />
+        public object Result { get; set; }
+
+        /// <inheritdoc />
+        public async Task Invoke()
+        {
+            // Execute each stage in the pipeline. Each successive call to this method will invoke the next stage.
+            // Stages which are not implemented (eg, because the user has not specified an interceptor) are skipped.
+            var numFilters = filters.Count;
+            if (stage < numFilters)
+            {
+                // Call each of the specified interceptors.
+                var systemWideFilter = this.filters[stage];
+                stage++;
+                await systemWideFilter.Invoke(this);
```

Comment:
when filters throw (either synchronously or return a faulted Task), will everything work fine? I couldn't find any coverage for that scenario (but tests are a little bit confusing, so there is a chance there is coverage and I'm missing it)